### PR TITLE
Lock Session Support

### PR DESCRIPTION
**Pull request by harshit-gangal**
Currently the lock function support relies on existing shard session within VTGate Session.
We will have a different Shard session called lock session to support this feature.

Locking and transaction are orthogonal and to not interfere with each other they need a separate code path to manage.

### CODE REVIEW

**Comment by harshit-gangal** [comment]
All the end to end test are related to lock are working. I have updated the correct expectation for end to end test when a tx failure while holding a lock.

Some unit test are pending for the new flow of ExecuteLock.

File: go/vt/vtgate/engine/lock.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/vindexes"
+)
+
+var _ Primitive = (*Lock)(nil)
+
+//Lock will mark the session as needing a
+//reserved connection and then execute the inner Primitive
```
Comment (systay) [inline go/vt/vtgate/engine/lock.go:31]: Comment needs to be updated

File: go/vt/vtgate/planbuilder/testdata/lock_cases.txt [hunk 2]
```diff
@@ -4,7 +4,13 @@
   "QueryType": "SELECT",
   "Original": "select get_lock('xyz', 10) from dual",
   "Instructions": {
-    "OperatorType": "Reserve"
+    "OperatorType": "Send",
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/lock_cases.txt:7]: I'm confused - shouldn't these be using the lock primitive?

File: go/vt/vtgate/engine/lock.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+func (l *Lock) GetFields(vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return nil, vterrors.New(vtrpc.Code_UNIMPLEMENTED, "not implements in lock primitive")
+}
+
+func (l *Lock) description() PrimitiveDescription {
+	other := map[string]interface{}{
+		"Query": l.Query,
+	}
+	return PrimitiveDescription{
+		OperatorType:      "Send",
```
Comment (systay) [inline go/vt/vtgate/engine/lock.go:98]: This is the wrong. It should be `Lock`

File: go/vt/vtgate/scatter_conn.go [hunk 4]
```diff
@@ -550,8 +552,82 @@ func (stc *ScatterConn) multiGoTransaction(
...
+	reservedID := info.reservedID
+
+	switch info.actionNeeded {
+	case nothing:
+		if reservedID == 0 {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "BUG: reservedID zero not expected %v", reservedID)
+		}
+		qr, err = qs.Execute(ctx, rs.Target, query.Sql, query.BindVariables, 0, reservedID, opts)
+	case reserve:
+		qr, reservedID, alias, err = qs.ReserveExecute(ctx, rs.Target, session.SetPreQueries(), query.Sql, query.BindVariables, 0, opts)
```
Comment (systay) [inline go/vt/vtgate/scatter_conn.go:607]: What is the `0` for? Can we add a comment?

**Comment by deepthi** [comment]
Review if anyone is looking for lock support in 7.0.

**Pull request merged**